cli_tools/gce_vm_image_import/importer/source.go

```diff
@@ -0,0 +1,198 @@
+			"file from bucket %q, file %q: %v", s.bucket, s.object, err)
+	}
+	defer rc.Close()
+
+	byteCountingReader := daisycommon.NewByteCountingReader(rc)
+	// Detect whether it's a compressed file by extracting compressed file header
+	if _, err = gzip.NewReader(byteCountingReader); err == nil {
+		return daisy.Errf("the input file is a gzip file, which is not supported by " +
+			"image import. To import a file that was exported from Google Compute " +
+			"Engine, please use image create. To import a file that was exported " +
+			"from a different system, decompress it and run image import on the " +
+			"disk image file directly")
+	}
+
+	// By calling gzip.NewReader above, a few bytes were read from the Reader in
+	// an attempt to decode the compression header. If the Reader represents
+	// an empty file, then BytesRead will be zero.
+	if byteCountingReader.BytesRead <= 0 {
+		return daisy.Errf("cannot import an image from an empty file")
+	}
+
+	return nil
+}
+
+// An importable source backed by a GCE disk image.
+type imageSource struct {
+	uri string
+}
+
+// Creates an imageSource from a reference to a GCE disk image. The syntax of the
+// reference is validated, but no I/O is performed to determine whether the image
+// exists or whether the calling user has access to it.
+func newImageSource(imagePath string) (resource, error) {
+	source := imageSource{
+		uri: param.GetGlobalResourcePath("images", imagePath),
+	}
+	return source, source.validate()
+}
+
+var imageNamePattern = regexp.MustCompile("^[a-z]([-a-z0-9]*[a-z0-9])?$")
+
+// Performs basic validation, focusing on error cases that
+// we've seen in the past.
+//
+// Two key offenders:
+//   1. Using a file path instead of an image name.
+//     Examples:
+//        - file://home/user/image.vmdk
+//        - gs://bucket/image.vmdk
+//        - gs://global/images/image
+//        - https://storage.googleapis.com/bucket/image
+//   2. Using a relative file path instead of an image name.
+//     Examples:
+//        - image.vmdk
+//        - path/to/image.vmdk
+//
+// This method does not validate whether the image exist, or whether
+// the calling user has access to it.
+func (s imageSource) validate() error {
+	parsed, err := url.Parse(s.uri)
+	if err != nil {
+		return err
+	}
+	if parsed.Scheme != "" {
+		return daisy.Errf(
+			"invalid image reference %q.", s.uri)
+	}
+
+	var imageName string
+	if slash := strings.LastIndex(s.uri, "/"); slash > -1 {
+		imageName = s.uri[slash+1:]
+	} else {
+		imageName = s.uri
+	}
+	if imageName == "" || len(imageName) > 63 {
+		return daisy.Errf(
+			"invalid image name %q. Image name must be 1-63 characters long, inclusive", imageName)
+	}
+	if !imageNamePattern.MatchString(imageName) {
+		return daisy.Errf(
+			"invalid image name %q. The first character must be a lowercase letter, and all "+
+				"following characters must be a dash, lowercase letter, or digit, except the last "+
```

Comment:
consider merging these logic to validation_utils.go

Reply:
With potentially-common code, my preference is to wait until it's needed in a second place, and then refactor it. I find that makes a cleaner API, since there are at least two concrete usages, rather than one concrete usage and a future envisioned usage. Also, sometimes that second usage never arises ..